File: backend/app/schemas/conformity.py
"""Conformity response schemas (Phase 5C)."""

import uuid
from datetime import date

from pydantic import BaseModel


class ConformityComponent(BaseModel):
    metric: str
    planned: float | None = None
    actual: float | None = None
    deviation_pct: float | None = None
    weight_used: float | None = None  # renormalised over scored components
    component_score: float | None = None  # 0..1, None when not comparable


class ConformityResult(BaseModel):
    conformity_pct: float | None = None
    classification: str | None = None
    components: list[ConformityComponent] = []
    status: str  # done | partial | missed | extra | pending | rest
    deviations: list[str] = []


class WeeklyConformity(BaseModel):
    week_number: int
    week_start: date
    week_end: date
    days_scored: int
    days_total: int
    pct: float | None = None
    by_sport: dict[str, float | None]


class PlanConformityResponse(BaseModel):
    plan_id: uuid.UUID
    overall_pct: float | None = None
    trend: str | None = None  # improving | declining | stable
    weeks: list[WeeklyConformity]
    patterns: list[str]


class DayConformityResponse(ConformityResult):
    plan_id: uuid.UUID
    day_id: uuid.UUID
    day_date: date
    sport: str
    planned_type: str | None = None


class LinkActivitiesResponse(BaseModel):
    linked: int

File: frontend/src/lib/api/types/conformity.ts
// ─── Plan Conformity (Phase 5C) ──────────────────────────────────────────

/** One weighted scoring component of a plan day (duration/power/TSS/…). */
export interface ConformityComponent {
  metric: string;
  planned: number | null;
  actual: number | null;
  deviation_pct: number | null;
  /** Weight applied to this component within the overall score (0–1). */
  weight_used: number;
  /** Per-component score (0–1). */
  component_score: number | null;
}

export type DayConformityStatus =
  | 'done'
  | 'partial'
  | 'missed'
  | 'extra'
  | 'pending'
  | 'rest';

/** GET /training-plans/{id}/days/{day_id}/conformity */
export interface DayConformityResponse {
  plan_id: string;
  day_id: string;
  day_date: string;
  sport: string;
  planned_type: string;
  conformity_pct: number | null;
  classification: string | null;
  components: ConformityComponent[];
  status: DayConformityStatus;
  deviations: string[];
}

/** One week's aggregate adherence inside a plan-conformity response. */
export interface WeekConformity {
  week_number: number;
  week_start: string;
  week_end: string;
  days_scored: number;
  days_total: number;
  pct: number | null;
  by_sport: Record<string, number>;
}

/** GET /training-plans/{id}/conformity */
export interface PlanConformityResponse {
  plan_id: string;
  overall_pct: number | null;
  trend: 'improving' | 'declining' | 'stable' | null;
  weeks: WeekConformity[];
  patterns: string[];
}

/** POST /training-plans/{id}/link-activities */
export interface LinkActivitiesResponse {
  linked: number;
}
